pygbm/splitting.py

```diff
@@ -192,38 +201,48 @@ def split_indices(context, split_info, sample_indices):
-        if context.constant_hessian:
-            ordered_hessians = context.all_hessians
-            for i, sample_idx in enumerate(sample_indices):
-                ordered_gradients[i] = context.all_gradients[sample_idx]
-        else:
-            ordered_hessians = np.empty_like(sample_indices,
-                                             dtype=loss_dtype)
-            for i, sample_idx in enumerate(sample_indices):
-                ordered_gradients[i] = context.all_gradients[sample_idx]
-                ordered_hessians[i] = context.all_hessians[sample_idx]
+    ctx = context  # shorter name to avoid various line breaks
+    n_samples = sample_indices.shape[0]
 
-    sum_gradients = ordered_gradients.sum()
-    if context.constant_hessian:
-        n_samples = sample_indices.shape[0]
-        sum_hessians = context.constant_hessian_value * float32(n_samples)
+    # Need to declare local variables, else they're not updated
+    # (see numba issue 3459)
+    ordered_gradients = ctx.ordered_gradients
+    ordered_hessians = ctx.ordered_hessians
+
+    # Populate ordered_gradients and ordered_hessians. (Already done for root)
+    # This is a parallelized version of the following vanilla code:
+    # for i range(n_samples):
+    #     ctx.ordered_gradients[i] = ctx.all_gradients[samples_indices[i]]
+    if sample_indices.shape[0] != ctx.all_gradients.shape[0]:
+        n_threads = numba.config.NUMBA_DEFAULT_NUM_THREADS
+        # Each threads writes data in ordered_xx from starts[thread_idx] to
+        # starts[thread_idx] + sizes[thread_idx]
+        sizes = np.full(n_threads, n_samples // n_threads, dtype=np.int32)
+        sizes[:n_samples % n_threads] += 1
+        starts = np.zeros(n_threads, dtype=np.int32)
+        starts[1:] = np.cumsum(sizes[:-1])
+        if ctx.constant_hessian:
+            for thread_idx in prange(n_threads):
+                for i in range(starts[thread_idx],
+                               starts[thread_idx] + sizes[thread_idx]):
+                    ordered_gradients[i] = ctx.all_gradients[sample_indices[i]]
+        else:
+            for thread_idx in prange(n_threads):
+                for i in range(starts[thread_idx],
+                               starts[thread_idx] + sizes[thread_idx]):
+                    ordered_gradients[i] = ctx.all_gradients[sample_indices[i]]
+                    ordered_hessians[i] = ctx.all_hessians[sample_indices[i]]
+
+    ctx.sum_gradients = ctx.ordered_gradients[:n_samples].sum()
+    if ctx.constant_hessian:
+        ctx.sum_hessians = ctx.constant_hessian_value * float32(n_samples)
     else:
-        sum_hessians = ordered_hessians.sum()
+        ctx.sum_hessians = ctx.ordered_hessians[:n_samples].sum()
 
-    return _parallel_find_split(
-        context, sample_indices, ordered_gradients, ordered_hessians,
-        sum_gradients, sum_hessians)
+    return _parallel_find_split(ctx, sample_indices)
```

Comment:
Because `_parallel_find_split` is a very short function, I think we could collapse it here. We could reuse the docstring of `_parallel_find_split` directly for the `find_node_split` function that has no docstring.

I believe collapsing the functions together would make the code easier to follow/understand.